pkg/operators/combiner/combiner.go

```diff
@@ -0,0 +1,309 @@
+// Copyright 2024 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package combiner is a data operator that combines data from a same data
+// source coming from different targets into a single data source. This is
+// useful when we run the same gadget in a distributed environment like
+// Kubernetes and we want to combine the data from all the nodes on the client
+// side so that we can perform further operations on the combined data, e.g.
+// sorting. Notice that this operator is useful only when we have data sources
+// of type array.
+package combiner
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	"google.golang.org/protobuf/proto"
+
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/datasource"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-service/api"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/operators"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/params"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/runtime"
+)
+
+const (
+	// We need it to be the first operator to run in client side. Only operators
+	// registering data sources should run before the combiner operator.
+	Priority             = -500
+	DataSourcePrefix     = "combined"
+	OperatorName         = "Combiner"
+	OperatorInstanceName = "CombinerInstance"
+)
+
+type combinerOperator struct{}
+
+func (o *combinerOperator) Name() string {
+	return OperatorName
+}
+
+func (o *combinerOperator) Init(params *params.Params) error {
+	return nil
+}
+
+func (o *combinerOperator) GlobalParams() api.Params {
+	return nil
+}
+
+func (o *combinerOperator) InstanceParams() api.Params {
+	return nil
+}
+
+func (o *combinerOperator) InstantiateDataOperator(gadgetCtx operators.GadgetContext, paramValues api.ParamValues) (operators.DataOperatorInstance, error) {
+	activate := false
+
+	targetsVar, ok := gadgetCtx.GetVar(runtime.RunTargets)
+	if !ok {
+		gadgetCtx.Logger().Debugf("No targets found. Skipping combiner operator")
+		return nil, nil
+	}
+
+	targets, ok := targetsVar.(int)
+	if !ok {
+		return nil, fmt.Errorf("invalid type for number of targets: %T, expected \"int\"", targetsVar)
+	}
+	if targets > 0 {
+		for _, ds := range gadgetCtx.GetDataSources() {
+			if ds.Type() == datasource.TypeArray {
```

Comment:
Should this be more explicit by using an annotation on the datasource?

`combiner.enable: true` on the data source; or if we stick to enabling by default allowing to do a `combiner.disable: true`.

Reply:
Also, in here you could already collect the datasources and pre-fill a struct (if transformed to `map[datasource.DataSource]combinerConfig`, where combinerConfig also contains the `chan datasource.PacketArray`, replacing the `pktBuffers`), so in `PreStart()` you could iterate over that instead of getting all datasources again and applying filters (is array, maybe annotations valid, etc).

Reply:
> > Should this be more explicit by using an annotation on the datasource?
> 
> I don't see a use case for a datasource that generates arrays that doesn't want to combine the output from all targets, especially due to the way we print them in the CLI (clean up screen each time a new batch arrives). Do you?

If you're just digesting the arrays as they come in from your nodes to do post-processing in your own third-party app (consuming just JSON for example), you might want to have separated events per node or so. I would even say that the default should be not to combine them as combining could be seen as "just" an optimization for certain use cases.

> > Also, in here you could already collect the datasources and pre-fill a struct
> 
> I didn't want to do that because `InstantiateDataOperator` is also called in `GetGadgetInfo`. Thus, it's a waste of resources just to then be discarded. I think we need to improve that first. WDYT?

I wouldn't worry too much about those resources tbh - it's not really much compared to everything else that is happening for `GetGadgetInfo()` anyway (parsing eBPF etc.). Also, since `GetGadgetInfo()` is only called on one node and can potentially be cached, it's called much less than `RunGadget()`, where the data isn't thrown away.

I think given that, I would prefer to have the logic of filtering/error checking only once and at the earliest possible time.

Reply:
Fair. I'll keep it enabled by default for now and will work on the possibility of disabling it through annotations in a follow-up PR.